src/librustc_mir/monomorphize/polymorphize.rs

```diff
@@ -0,0 +1,283 @@
+        }
+
+        // Consider every non-generic parameter used by default.
+        match param.kind {
+            ty::GenericParamDefKind::Type { .. } => (),
+            _ => {
+                used_parameters.insert(param.index);
+            }
+        }
+    }
+
+    if let Some(parent) = generics.parent {
+        mark_used_by_default_parameters(
+            tcx,
+            def_id,
+            tcx.generics_of(parent),
+            false,
+            used_parameters,
+        );
+    }
+}
+
+/// Emit an error for the function represented by `def_id`, labelling each generic parameter which
+/// was unused.
+fn emit_used_generic_parameters_error<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    generics: &'tcx ty::Generics,
+    used_parameters: &BitSet<u32>,
+) {
+    debug!("emit_used_generic_parameters_error: used_parameters={:?}", used_parameters);
+    let fn_span = match tcx.opt_item_name(def_id) {
+        Some(ident) => ident.span,
+        _ => tcx.def_span(def_id),
+    };
+
+    let mut err = tcx.sess.struct_span_err(fn_span, "item has unused generic parameters");
+
+    let mut unused_parameters = BitSet::new_filled(used_parameters.domain_size());
+    unused_parameters.subtract(used_parameters);
+    debug!("emit_used_generic_parameters_error: unused_parameters={:?}", unused_parameters);
+
+    for index in unused_parameters.iter() {
+        let param = ty::ParamTy::new(index, Symbol::intern(&format!("P{}", index)));
+        // Can safely call `type_param` as all non-type parameters are considered used.
+        let generic_param_def = generics.type_param(&param, tcx);
+        debug!(
+            "emit_used_generic_parameters_error: generic_param_def={:?} param={:?}",
+            generic_param_def, param
+        );
+        let def_span = tcx.def_span(generic_param_def.def_id);
+        err.span_label(
+            def_span,
+            &format!("generic parameter `{}` is unused", generic_param_def.name),
+        );
+    }
+
+    err.emit();
+}
+
+/// Visitor used to aggregate generic parameter uses.
+struct UsedGenericParametersVisitor<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    body: ReadOnlyBodyAndCache<'a, 'tcx>,
+    used_parameters: &'a mut BitSet<u32>,
+}
+
+impl<'a, 'tcx> UsedGenericParametersVisitor<'a, 'tcx> {
+    fn visit_foldable<T: std::fmt::Debug + TypeFoldable<'tcx>>(&mut self, ty: T) {
+        debug!("visit_foldable: ty={:?}", ty);
+        // Exit early if this type doesn't use any type parameters.
+        if !ty.has_param_types() {
+            return;
+        }
+
+        // Use `TypeVisitor` to find any type parameters.
+        ty.visit_with(self);
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for UsedGenericParametersVisitor<'a, 'tcx> {
+    fn visit_local_decl(&mut self, local: Local, local_decl: &LocalDecl<'tcx>) {
+        debug!("visit_local_decl: local_decl={:?}", local_decl);
+        let ty = local_decl
+            .ty
+            .builtin_deref(false)
```

Comment:
Is this for the closure thing? A better way to check is that if `self.def_id` is a closure, you can ignore `Local::new(1)` (the first argument).

Reply:
I've fixed this in dd24b5b.